ldap/servers/slapd/auditlog.c

```diff
@@ -228,6 +228,7 @@ add_entry_attrs(Slapi_Entry *entry, lenstr *l)
                 addlenstr(l, log_val);
                 addlenstr(l, "\n");
             }
+            charray_free(vals);
```

Comment:
This might be overkill, but I really  think we could benefit from looping over the entry attr array by reference, rather than copying/freeing the values.  I did something similar with slapi_entry_attr_get_ref(), and we could create a similar function for the attr array.   

If we can we should always avoid mallcoing/freeing if possible.  It will help with memory usage/fragmentation.

Thoughts?

Reply:
Good point and it will also avoid lock contention to access the heap.
   but if we want to avoid any malloc IMHO we have better to iterate directly on the value berval's and then also avoid the strlen call. It should looks like:  
  ```
        slapi_entry_attr_find(e, type, &attr);
        int i = slapi_attr_first_value(attr, &sval);
        while (!r && i != -1) {
            const struct berval *val = slapi_value_get_berval(sval);
```

Reply:
Sounds great, and perhaps this approach can be used else where in the code (future ticket).  

The GCC team says how DS uses memory is "unusual" and it causes a really large memory footprint.   This is why we use jemalloc today instead of gcc.  Part of this is due to all these small and frequent allocations the server does.  It's been an ongoing silent effort to slowly remove these mallocs/frees where we can.